### PR TITLE
hostmux: init at 1.4.0

### PR DESCRIPTION
**Pull request by fernsehmuell**
## Description of changes
Small wrapper script for tmux to easily connect to a series of hosts via ssh and open a split pane for each of the hosts.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: maintainers/maintainer-list.nix [hunk 1]
```diff
@@ -5488,6 +5488,13 @@
       }
     ];
   };
+  fernsehmuell = {
+    email = "fernsehmuel@googlemail.com";
+    matrix = "@fernsehmuell:matrix.org";
+    github = "fernsehmuell";
+    githubId = "5198058";
```
Comment (antonmosich) [inline maintainers/maintainer-list.nix:5495]: Your `githubId` should not be in quotes

Comment (fernsehmuell) [reply maintainers/maintainer-list.nix:5495]: Oh i see, will fix!

**Comment by antonmosich** [comment]
You'll need to work a little more on your PR. Currently, when I try to run this, I get an error, that tmux isn't found. If I hadn't installed ssh on my system I would get the same error. You need to edit your derivation such that calls to tmux get replaced by absolute paths to the tmux binary in the nix store. The same needs to be done for ssh. You should be able to use the substituteInPlace function for that.
You _should_ also add `meta.mainProgram` as `lib.getExe` will no longer assume the package name to be the name of the main program.
Also, in order to be able to use the added program properly, you should also include it in `all-packages.nix`.
Read the [nixpkgs manual](https://nixos.org/manual/nixpkgs/stable) for more information.

File: maintainers/maintainer-list.nix [hunk 2]
```diff
@@ -5488,6 +5488,13 @@
       }
     ];
   };
+  fernsehmuell = {
```
Comment (AndersonTorres) [inline maintainers/maintainer-list.nix:5491]: Split this in a separate commit  (**no need to open a PR**).

File: pkgs/tools/misc/hostmux/default.nix [hunk 3]
```diff
@@ -0,0 +1,33 @@
+{ lib, stdenv, fetchFromGitHub, installShellFiles, openssh, tmux, ... }:
```
Comment (AndersonTorres) [inline pkgs/tools/misc/hostmux/default.nix:1]: Do not use ellipsis.
```suggestion
{ lib
, stdenv
, fetchFromGitHub
, installShellFiles
, openssh
, tmux
}:
```

File: pkgs/tools/misc/hostmux/default.nix [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+
+stdenv.mkDerivation rec {
```
Comment (AndersonTorres) [inline pkgs/tools/misc/hostmux/default.nix:3]:
```suggestion
stdenv.mkDerivation (finalAttrs: {
```

File: pkgs/tools/misc/hostmux/default.nix [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+  version = "1.4.0";
+  pname = "hostmux";
```
Comment (AndersonTorres) [inline pkgs/tools/misc/hostmux/default.nix:5]:
```suggestion
  pname = "hostmux";
  version = "1.4.0";
```

File: pkgs/tools/misc/hostmux/default.nix [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+
+  nativeBuildInputs = [ installShellFiles ];
+
+  buildInputs = [ openssh tmux ];
+
```
Comment (AndersonTorres) [inline pkgs/tools/misc/hostmux/default.nix:10]:
```suggestion
```

File: pkgs/tools/misc/hostmux/default.nix [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+  src = fetchFromGitHub {
+    owner = "hukl";
+    repo = pname;
```
Comment (AndersonTorres) [inline pkgs/tools/misc/hostmux/default.nix:13]:
```suggestion
    repo = "hostmux";
```

File: pkgs/tools/misc/hostmux/default.nix [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+    rev = version;
+    sha256 = "0h1wcbzlckjzg0y2lljf840kxvqzv5c6snn5bcbcpp0lbd07plx1";
+  };
+
```
Comment (AndersonTorres) [inline pkgs/tools/misc/hostmux/default.nix:17]:
```suggestion

  nativeBuildInputs = [ 
    installShellFiles
  ];

  buildInputs = [
    openssh
    tmux 
  ];

```

File: pkgs/tools/misc/hostmux/default.nix [hunk 9]
```diff
@@ -0,0 +1,33 @@
...
+  nativeBuildInputs = [ installShellFiles ];
+
+  buildInputs = [ openssh tmux ];
+
+  src = fetchFromGitHub {
+    owner = "hukl";
+    repo = pname;
+    rev = version;
+    sha256 = "0h1wcbzlckjzg0y2lljf840kxvqzv5c6snn5bcbcpp0lbd07plx1";
```
Comment (AndersonTorres) [inline pkgs/tools/misc/hostmux/default.nix:15]: Please use SRI hash.

```suggestion
    hash = "sha256-odN7QFsU3MsWW8VabVjZH+8+AUFOUio8eF9ORv9iPEA=";
```

File: pkgs/tools/misc/hostmux/default.nix [hunk 10]
```diff
@@ -0,0 +1,33 @@
...
+  pname = "hostmux";
+
+  nativeBuildInputs = [ installShellFiles ];
+
+  buildInputs = [ openssh tmux ];
+
+  src = fetchFromGitHub {
+    owner = "hukl";
+    repo = pname;
+    rev = version;
```
Comment (AndersonTorres) [inline pkgs/tools/misc/hostmux/default.nix:14]:
```suggestion
    rev = finalAttrs.version;
```

File: pkgs/tools/misc/hostmux/default.nix [hunk 11]
```diff
@@ -0,0 +1,33 @@
...
+    sha256 = "0h1wcbzlckjzg0y2lljf840kxvqzv5c6snn5bcbcpp0lbd07plx1";
+  };
+
+  installPhase = ''
+    install -Dm755 hostmux $out/bin/hostmux
+    installManPage man/hostmux.1
+    installShellCompletion --zsh zsh-completion/_hostmux
+  '';
```
Comment (AndersonTorres) [inline pkgs/tools/misc/hostmux/default.nix:22]:
```suggestion
  installPhase = ''
    runHook preInstall
    
    install -Dm755 hostmux $out/bin/hostmux
    installManPage man/hostmux.1
    installShellCompletion --zsh zsh-completion/_hostmux

    runHook postInstall
  '';
```

File: pkgs/tools/misc/hostmux/default.nix [hunk 12]
```diff
@@ -0,0 +1,33 @@
...
+
+  meta = with lib; {
+    description =
+      "Small wrapper script for tmux to easily connect to a series of hosts via ssh and open a split pane for each of the hosts.";
+    downloadPage = "https://github.com/hukl/hostmux/releases/";
+    homepage = "https://github.com/hukl/hostmux";
+    license = licenses.mit;
+    maintainers = with maintainers; [ fernsehmuell ];
+    platforms = with platforms; linux ++ darwin;
+  };
```
Comment (AndersonTorres) [inline pkgs/tools/misc/hostmux/default.nix:32]:
```suggestion
  meta = {
    description = "Small wrapper script for tmux to easily connect to a series of hosts via ssh and open a split pane for each of the hosts";
    homepage = "https://github.com/hukl/hostmux";
    license = lib.licenses.mit;
    maintainers = with lib.maintainers; [ fernsehmuell ];
    platforms = lib.platforms.unix;
  };
```

File: pkgs/tools/misc/hostmux/default.nix [hunk 13]
```diff
@@ -0,0 +1,33 @@
...
+}
```
Comment (AndersonTorres) [inline pkgs/tools/misc/hostmux/default.nix:33]:
```suggestion
})
```

**Review by AndersonTorres — CHANGES_REQUESTED** [review]

**Comment by fernsehmuell** [comment]
@AndersonTorres thanks for your time and your review. I applied all you suggestions.

**Comment by fernsehmuell** [comment]
@antonmosich I thought buildInputs = [ tmux openssh ] solves the problem that ssh + tmux are found, do i really have to patch the shell script?

**Comment by antonmosich** [comment]
Yes, you still need to patch the script.
You can try out `nix-shell -E 'with import /path/to/your/nixpkgs { }; mkShell { buildInputs = [ hostmux ]; }' --pure`. The `--pure` makes sure that there are no other environment variables "polluting" your shell (such as a pre-populated PATH).
Running hostmux in that shell should fail.

File: pkgs/tools/misc/hostmux/default.nix [hunk 14]
```diff
@@ -0,0 +1,58 @@
...
+
+  postPatch = ''
+    substituteInPlace hostmux \
+      --replace "SSH_CMD=ssh" "SSH_CMD=${openssh}/bin/ssh" \
+      --replace "tmux -2" "${tmux}/bin/tmux -2" \
+      --replace "tmux split-window" "${tmux}/bin/tmux split-window" \
+      --replace "tmux select-layout" "${tmux}/bin/tmux select-layout" \
+      --replace "tmux show-window-option" "${tmux}/bin/tmux show-window-option" \
+      --replace "tmux set-window-option" "${tmux}/bin/tmux set-window-option" \
+      --replace "tmux send-keys" "${tmux}/bin/tmux send-keys"
```
Comment (antonmosich) [inline pkgs/tools/misc/hostmux/default.nix:47]:
```suggestion
      --replace "tmux " "${tmux}/bin/tmux "
```
Note the space after tmux.
I believe that should suffice.

**Review by antonmosich — COMMENTED** [review]
You can squash the two latest commits - you don't need a separate one for adding you package to all-packages.
Please also reword your commit message on the first commit to be like `maintainers: add fernsehmuell`.

File: pkgs/tools/misc/hostmux/default.nix (hunk 14 above)
Comment (fernsehmuell) [reply pkgs/tools/misc/hostmux/default.nix:47]: no this would also replace "USAGE= hostmux -h" but i found a smaller version with "tmux s" and "tmux -2"

Comment (antonmosich) [reply pkgs/tools/misc/hostmux/default.nix]: Thanks I didn't spot this!

**Review by antonmosich — APPROVED** [review]

File: pkgs/tools/misc/hostmux/default.nix [hunk 15]
```diff
@@ -0,0 +1,54 @@
...
+  nativeBuildInputs = [
+    installShellFiles
+  ];
+
+  buildInputs = [
+    openssh
+    tmux
+  ];
+
+  installPhase = ''
```
Comment (AndersonTorres) [inline pkgs/tools/misc/hostmux/default.nix:29]: installPhase should be after postPatch

Comment (fernsehmuell) [reply pkgs/tools/misc/hostmux/default.nix:36]: fixed it

**Review by AndersonTorres — APPROVED** [review]
